### PR TITLE
add `title` and `capitalize` methods on type Str

### PR DESCRIPTION
**Pull request by slashformotion**
I am a rust newbie, i have no idea what i am doing. 
I will add docs later

### CODE REVIEW

File: crates/typst/src/eval/str.rs [hunk 1]
```diff
@@ -346,6 +346,27 @@ impl Str {
...
+        words.join(" ").into()
+    }
+
+    // Capitalize the first letter.
+    pub fn capitalize(&self) -> Self {
+        uppercase_first_letter(self.0.as_str()).into()
+    }
+}
+
+// uppercase the first letter of an str
```
Comment (andreasKroepelin) [inline crates/typst/src/eval/str.rs:363]: an str --> a str

File: crates/typst/src/eval/str.rs [hunk 2]
```diff
@@ -346,6 +346,27 @@ impl Str {
         self.locate_opt(index)?
             .ok_or_else(|| out_of_bounds(index, self.len()))
     }
+
+    // Capitalize the first letter of every word.
```
Comment (bluebear94) [inline crates/typst/src/eval/str.rs:350]: These should be doc-comments (three leading slashes) and be in third person: `/// Capitalizes the first letter of every word.`

**Comment by laurmaedje** [comment]
Title case doesn't just uppercase the first word of each letter as far as I know. There are also exceptions for certain words. I don't know how it works / if it's a thing for languages other than English. Furthermore, I think `lower`, `upper` and this should be consistent. For once, both should work on content and then there is the topic of method vs function. It's intricate.

**Comment by slashformotion** [comment]
I wanted to reproduce the `.title` method of python strings (see [docs](https://docs.python.org/3/library/stdtypes.html#str.title)). Seems that if capitalise every word

image


a test with arbitrary strings 
image

When it comes to content, i failed to implement the methods (still a rust noob).

**Comment by slashformotion** [comment]
I am closing it for now since I don't have the time to work on it

**Comment by slashformotion** [comment]
I can do another PR with only capitalize if you wish @laurmaedje .

**Comment by freundTech** [comment]
The problem with title case is that it's not really standardized.
Title case basically capitalized the first letter of each word, except for minor words. What counts as a minor word varies between the style guides of different institutions and publishers though.

I found this site, which has a title case converter for multiple styles and links to it's sources: https://capitalizemytitle.com/

**Comment by SylvanFranklin** [comment]
Any progress on this? It seems a shame to import an entire package for basic capitalization functionality on a string.

**Comment by alerque** [comment]
Capitalization of a string is not exactly as "basic" as some might expect.

If you are interested I have a working [prototype of decasify](https://github.com/alerque/decasify/pull/35) as a Typst package that can handle capitalization of strings (titles, sentences, etc.) in titlecase, sentence case, etc. for several languages. I was still fiddling with the Typst interface functions but it's pretty close to being ready for submission as a package.

**Comment by SylvanFranklin** [comment]
You are right this problem is a lot deeper than it appears because of different language conventions and what count's as a minor word, as others have already mentioned above. I was just googling for a way to capitalize a single English word. I came up with: 
```typ
let w = "word"
w = upper(w.at(0)) + w.slice(1, w.len())
```

**Comment by slashformotion** [comment]
> Any progress on this? It seems a shame to import an entire package for basic capitalization functionality on a string.

nodejs vibes

**Pull request closed without merging**